### DISCUSSION TITLE
Problem does not follow DCP rules, while it does?

### DISCUSSION THREAD

**Discussion by siyuanpurdue** [cvxpy/cvxpy#2398] [Q&A]
I have an expression of this form: x[i] * cp.abs((lot+tau*all1)@v0). It is of type Expression(UNKNOWN, UNKNOWN, (2, 6)), which is surprising to me since I have follows.

x = cp.Variable(numNodes+1)

obj_constraints = []
 # I have constraints saying x are all non-negative.
for i in range(numNodes+1):
        obj_constraints += [x[i] >= 0]

However, cp.abs((lot+tau * all1 ) @ v0) is Expression(CONVEX, NONNEGATIVE, (2, 6)). 

x[i] is Expression(AFFINE, UNKNOWN, ())?? I thought x[i] should be non-negative so that the overall expression  x[i] * cp.abs((lot+tau*all1)@v0) would be judged convex? 

If I want to follow DCP rules, what should I do/formulate my expression/problem?

**Comment by siyuanpurdue** [comment]
Below is the complete error message:

DCPError: Problem does not follow DCP rules. Specifically:
The following constraints are not DCP:
reshape([[ 1.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  0.  0.  0.]
 [ 0.  0.  1.  0.  0.  0.]
 [ 0.  0.  0.  1.  0.  0.]
 [ 0.  0.  0.  0.  1.  0.]
 [ 0.  0.  0.  0.  0.  1.]
 [-1.  0.  0.  0.  0.  0.]
 [ 0. -1.  0.  0.  0.  0.]
 [ 0.  0. -1.  0.  0.  0.]
 [ 0.  0.  0. -1.  0.  0.]
 [ 0.  0.  0.  0. -1.  0.]
 [ 0.  0.  0.  0.  0. -1.]].T @ var2539[0, 0:12], (6, 1), F) == [[1. 0.]] @ abs(([[1. 0.]
 [0. 1.]] + -[[0.00015 0.00015]
 [0.00015 0.0003 ]] @ diag_vec(reshape([[1. 0. 0.]
 [0. 1. 0.]] @ var2538, (2,), F))) @ [[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]]) + [[0.1 0.1]] @ abs([[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]]).T , because the following subexpressions are not:
|--  reshape([[ 1.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  0.  0.  0.]
 [ 0.  0.  1.  0.  0.  0.]
 [ 0.  0.  0.  1.  0.  0.]
 [ 0.  0.  0.  0.  1.  0.]
 [ 0.  0.  0.  0.  0.  1.]
 [-1.  0.  0.  0.  0.  0.]
 [ 0. -1.  0.  0.  0.  0.]
 [ 0.  0. -1.  0.  0.  0.]
 [ 0.  0.  0. -1.  0.  0.]
 [ 0.  0.  0.  0. -1.  0.]
 [ 0.  0.  0.  0.  0. -1.]].T @ var2539[0, 0:12], (6, 1), F) == [[1. 0.]] @ abs(([[1. 0.]
 [0. 1.]] + -[[0.00015 0.00015]
 [0.00015 0.0003 ]] @ diag_vec(reshape([[1. 0. 0.]
 [0. 1. 0.]] @ var2538, (2,), F))) @ [[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]]) + [[0.1 0.1]] @ abs([[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]]).T
reshape([[ 1.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  0.  0.  0.]
 [ 0.  0.  1.  0.  0.  0.]
 [ 0.  0.  0.  1.  0.  0.]
 [ 0.  0.  0.  0.  1.  0.]
 [ 0.  0.  0.  0.  0.  1.]
 [-1.  0.  0.  0.  0.  0.]
 [ 0. -1.  0.  0.  0.  0.]
 [ 0.  0. -1.  0.  0.  0.]
 [ 0.  0.  0. -1.  0.  0.]
 [ 0.  0.  0.  0. -1.  0.]
 [ 0.  0.  0.  0.  0. -1.]].T @ var2540[0, 0:12], (6, 1), F) == Promote(var2538[0], (2, 6)) @ abs(([[1. 0.]
 [0. 1.]] + -[[0.00015 0.00015]
 [0.00015 0.0003 ]] @ diag_vec(reshape([[1. 0. 0.]
 [0. 1. 0.]] @ var2538, (2,), F)) + [[0.1 0.1]
 [0.1 0.1]]) @ [[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]]).T , because the following subexpressions are not:
|--  Promote(var2538[0], (2, 6)) @ abs(([[1. 0.]
 [0. 1.]] + -[[0.00015 0.00015]
 [0.00015 0.0003 ]] @ diag_vec(reshape([[1. 0. 0.]
 [0. 1. 0.]] @ var2538, (2,), F)) + [[0.1 0.1]
 [0.1 0.1]]) @ [[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]])
reshape([[ 1.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  0.  0.  0.]
 [ 0.  0.  1.  0.  0.  0.]
 [ 0.  0.  0.  1.  0.  0.]
 [ 0.  0.  0.  0.  1.  0.]
 [ 0.  0.  0.  0.  0.  1.]
 [-1.  0.  0.  0.  0.  0.]
 [ 0. -1.  0.  0.  0.  0.]
 [ 0.  0. -1.  0.  0.  0.]
 [ 0.  0.  0. -1.  0.  0.]
 [ 0.  0.  0.  0. -1.  0.]
 [ 0.  0.  0.  0.  0. -1.]].T @ var2539[1, 0:12], (6, 1), F) == [[0. 1.]] @ abs(([[1. 0.]
 [0. 1.]] + -[[0.00015 0.00015]
 [0.00015 0.0003 ]] @ diag_vec(reshape([[1. 0. 0.]
 [0. 1. 0.]] @ var2538, (2,), F))) @ [[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]]) + [[0.1 0.1]] @ abs([[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]]).T , because the following subexpressions are not:
|--  reshape([[ 1.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  0.  0.  0.]
 [ 0.  0.  1.  0.  0.  0.]
 [ 0.  0.  0.  1.  0.  0.]
 [ 0.  0.  0.  0.  1.  0.]
 [ 0.  0.  0.  0.  0.  1.]
 [-1.  0.  0.  0.  0.  0.]
 [ 0. -1.  0.  0.  0.  0.]
 [ 0.  0. -1.  0.  0.  0.]
 [ 0.  0.  0. -1.  0.  0.]
 [ 0.  0.  0.  0. -1.  0.]
 [ 0.  0.  0.  0.  0. -1.]].T @ var2539[1, 0:12], (6, 1), F) == [[0. 1.]] @ abs(([[1. 0.]
 [0. 1.]] + -[[0.00015 0.00015]
 [0.00015 0.0003 ]] @ diag_vec(reshape([[1. 0. 0.]
 [0. 1. 0.]] @ var2538, (2,), F))) @ [[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]]) + [[0.1 0.1]] @ abs([[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]]).T
reshape([[ 1.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  0.  0.  0.]
 [ 0.  0.  1.  0.  0.  0.]
 [ 0.  0.  0.  1.  0.  0.]
 [ 0.  0.  0.  0.  1.  0.]
 [ 0.  0.  0.  0.  0.  1.]
 [-1.  0.  0.  0.  0.  0.]
 [ 0. -1.  0.  0.  0.  0.]
 [ 0.  0. -1.  0.  0.  0.]
 [ 0.  0.  0. -1.  0.  0.]
 [ 0.  0.  0.  0. -1.  0.]
 [ 0.  0.  0.  0.  0. -1.]].T @ var2540[1, 0:12], (6, 1), F) == Promote(var2538[1], (2, 6)) @ abs(([[1. 0.]
 [0. 1.]] + -[[0.00015 0.00015]
 [0.00015 0.0003 ]] @ diag_vec(reshape([[1. 0. 0.]
 [0. 1. 0.]] @ var2538, (2,), F)) + [[0.1 0.1]
 [0.1 0.1]]) @ [[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]]).T , because the following subexpressions are not:
|--  Promote(var2538[1], (2, 6)) @ abs(([[1. 0.]
 [0. 1.]] + -[[0.00015 0.00015]
 [0.00015 0.0003 ]] @ diag_vec(reshape([[1. 0. 0.]
 [0. 1. 0.]] @ var2538, (2,), F)) + [[0.1 0.1]
 [0.1 0.1]]) @ [[ 0.00015  0.00015 -0.00015 -0.00015 -0.00015 -0.00015]
 [ 0.00015  0.0003  -0.00015 -0.0003  -0.00015 -0.0003 ]])

**Comment by Transurgeon** [comment]
Hi @siyuanpurdue you can specify non-negativity constraints directly when you initialise your variables.
```
x = cp.Variable(numNodes+1, nonneg=True)
```
This should fix your issue :).

**Reply by siyuanpurdue** [reply]
Yes, I tried this and it did not work... @Transurgeon
